### DISCUSSION TITLE
Searchable: External table with name <...> does not exist in connection <...>

### DISCUSSION THREAD

**Discussion by ash-peng** [thoughtspot/cs_tools#164] [Q&A] [answered]
Hi,

We are encountering some errors and looking for help on solving them. The tool is Searchable deploy. It is giving out errors including "Tables do not exist" and "external with name... does not exist in connection".

The tables that do not exist are:

TS_GROUP_MEMBERSHIP
TS_ORG
VW_TS_BI_SERVER_ADVANCED_AUTHOR
TS_METADATA_OBJECT
TS_DEPENDENT_OBJECT
TS_CLUSTER
TS_DATA_SOURCE
TS_USER
TS_BI_SERVER
TS_GROUP
TS_METADATA_COLUMN

---
The external tables that do not exist are:

<database>.<schema>.ts_bi_server
<database>.<schema>.ts_cluster
<database>.<schema>.ts_column_synonym
<database>.<schema>.ts_data_source
<database>.<schema>.ts_dependent_object
<database>.<schema>.ts_group
<database>.<schema>.ts_xref_principal
<database>.<schema>.ts_group_privilege
<database>.<schema>.ts_metadata_column
<database>.<schema>.ts_metadata_object
<database>.<schema>.ts_org
<database>.<schema>.ts_xref_org
<database>.<schema>.ts_sharing_access
<database>.<schema>.ts_tag
<database>.<schema>.ts_tagged_object
<database>.<schema>.ts_user

---

When I log into ThoughtSpot in the browser with an admin account, all of the "TS: ..." titled items we have are worksheets and none of them are tables. An example would be "TS: Query Stats Worksheet".

Looking forward to your response, thanks!

**Comment by boonhapus** [comment]
Hey @ash-peng , have you run the commands `searchable metadata` and `searchable bi-server` to load the data to your database yet? These commands will also create the tables in the database as well.

`searchable deploy` can come last and only needs to be run once.

I would recommend using the latest version of CS Tools (v1.5.9) if you're on any dialect other than Snowflake.

**Reply by ash-peng** [reply]
Yes, both these commands have been run. Syncer is using CSV. The three commands have been run in order, in the same directory. CS Tools is in the newest version. What else are we doing wrong? Thanks

**Reply by boonhapus** [reply]
Apologies @ash-peng , I didn't see your reply. If you @boonhapus , I get a notification.

The `deploy` command uses the **[ThoughtSpot TML APIs](https://developers.thoughtspot.com/docs/restV2-playground?apiResourceId=http%2Fapi-endpoints%2Fmetadata%2Fimport-metadata-tml)** to load Tables, Worksheets, and Liveboards to your thoughtspot cluster. TML requires that you are working with a connected database like Snowflake, BigQuery, etc.

The `metadata` and `bi-server` commands will create tables in your external database and insert data into them via CS Tools's **[Syncer protocol](https://thoughtspot.github.io/cs_tools/syncer/what-is/)**. So you'll need to run these first with a database syncer, and then run the `deploy` command to load the TML to your ThoughtSpot cluster.

**Reply by ash-peng** [reply]
Thank you @boonhapus , I get it now. You're saying that we should try connecting to an external database like **Databricks**, and use it instead of CSV for the Syncer.
We will try that and report if there is any remaining issue. Thanks again!

**Reply by ash-peng** [reply]
Hi @boonhapus , does SQLite work?
For example, I have the metadata and bi-server data already inserted through the two commands in `.\data.db`.

I tried the following command but the error still is the same with what is in the orig post:
`cs_tools tools searchable deploy --connection-guid <connection-guid> --database data.db --schema <schema-name-in-thoughtspot> --config <config-name>`

Update: I tried different schema names and neither of "main" or "sqlite_schema" works.
I am sure that `data.db` contains all of the table names.

Update 2: I also tried `--database ThoughtSpot` which didn't work (like before). My question is: how can we let ThoughtSpot know to find the extra tables in the exported `data.db` file we have in the filesystem?

Thanks!

**Reply by boonhapus** [reply]
@ash-peng ThoughtSpot cannot connect to your fs. It works with databases like Snowflake, Databricks, BigQuery, etc. What database is currently connected to your ThoughtSpot cluster? You can find the Connections under the Data tab in the ThoughtSpot UI.

If you've deployed ThoughtSpot to your own VPC or hardware, then you could potentially use Falcon to the data to ThoughtSpot directly.. but this is a legacy connection mechanism.

**Reply by ash-peng** [reply]
@boonhapus I see now. So only Snowflake, Databricks, and...? We have some Databricks connections but we are not actively using them if this makes sense. We are not using any VPC (assuming it means Virtual Private Cloud) or hardware either. Please advise on the most convenient way to make CS Tools work in this circumstance... Thank you for the ongoing replies!

**Reply by boonhapus** [reply]
What connection type do you mainly use in ThoughtSpot @ash-peng ?

**Reply by boonhapus** [reply]
@ash-peng I'm a little confused 😅 what are you using ThoughtSpot for, if you don't have any connections and don't load data to Falcon?

**Reply by ash-peng** [reply]
@boonhapus  We are using Microsoft Azure Synapse for our main database. We have a databricks connection, but it's not actively used. I checked the Syncer page and didn't find Azure Synapse... is it not workable with CS Tools?

**Reply by boonhapus** [reply]
@ash-peng yeah that's correct. I'll check tomorrow to see if we have a test bench for Synapse at ThoughtSpot, to see if I can write a Syncer that supports MSSQL, but my guess is it would be easier for you to export using CSV, and use your own data loading process into Synapse. 

If you run the `searchable metadata` or `searchable bi-server` commands with `--syncer mock://` , you should be able print the DDL for each table that would be loaded to the database.

**Reply by ash-peng** [reply]
@boonhapus Hi, thank you for the explanation of the workaround! 

I chatted with my team lead today about all we've discussed, and their current preferred approach is to use Databricks - though our current data is in Synapse (we have Databricks connections that are unused). Could you verify (or not verify) that it is okay to have the actual ThoughtSpot data in one database and the metadata in another, because even though we are slowly moving ThoughtSpot data from Synapse to Databricks it will still take some time. Even if the data will be in Databricks it may end up in another schema/catalog. Thanks a bunch!

**Reply by boonhapus** [reply, marked as answer]
@ash-peng that all makes sense, good to hear about the move to dbx! It's a fantastic database.

CS Tools fetches information about ThoughtSpot metadata from the ThoughtSpot APIs. The data supporting those APIs actually resides in a ts-internal proprietary database that customers are not able to interact with directly.

There is no harm in storing the data that CS Tools fetches, in Databricks, and re-exposing it via its own Connection. This is actually the preferred method - creating a separate connection entirely.

**Reply by boonhapus** [reply]
If you would like, you can email me directly and include your lead(s) if it makes things easier for you to make the architecture decisions.

nicholas.cooper@thoughtspot.com

**Reply by ash-peng** [reply]
@boonhapus Thank you so much!!!
